### DISCUSSION TITLE
bug: `TRPCClient` from `createTRPCReact<RouterType>().createClient` is incorrectly typed

### DISCUSSION THREAD

**Discussion by jetaggart** [trpc/trpc#2926] [Q&A] [answered]
### Provide environment information

```

  System:
    OS: macOS 12.5.1
    CPU: (8) arm64 Apple M1
    Memory: 157.83 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.19.1 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Watchman: 2022.09.12.00 - /opt/homebrew/bin/watchman
  Browsers:
    Brave Browser: 106.1.44.105
    Chrome: 105.0.5195.125
    Firefox: 105.0.1
    Safari: 16.0
  npmPackages:
    @trpc/react: ^10.0.0-proxy-beta.15 => 10.0.0-proxy-beta.15
    react: 18.0.0 => 18.0.0 
    typescript: ~4.7.4 => 4.7.4 

```

### Describe the bug

If I create a trpc react proxy and client like:

```typescript
export const trpc = createTRPCReact<UserRouter>()
export const trpcClient: TRPCClient<UserRouter> = trpc.createClient({
  links: [
    httpBatchLink({
      url: "http://localhost:3000/user"
    }),
  ],
})
```

And try to use the `trpcClient` directly, it does not have the correct types:

```typescript
        const createAccountResult = await trpcClient.mutation("account.create")
```

Will return: `TS2345: Argument of type 'string' is not assignable to parameter of type 'never'.`

### Link to reproduction

https://stackblitz.com/github/trpc/examples-next-minimal-starter

### To reproduce

See above

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment]
You're using a deprecated API, aren't you?

**Reply by jetaggart** [reply]
I don't think so? I'm following the docs here: https://trpc.io/docs/v10/react and have the same code, `createReactQueryHooks` is the old way `createTRPCReact` is the new, correct?

To be clear, I'm using `trpc.account.create.useMutation` when I'm inside of a react component, where I can use hooks. However, I have some places (such as a redux saga, or a recoil atom) that I'd like to use the client directly where hooks are not allowed. I can do this through the client created from `createClient` but the typing is incorrect. I can ignore the type error explicitly and it works, like this:

```typescript
const createAccountResult = await trpcClient.mutation("account.create" as never)
```

Essentially, even though I have the hooks based client, there are places where I want the vanilla client. In a similar way I can use `QueryClient` outside the react for straight `react-query`

**Comment by chinanderm** [comment]
I had this same issue crop up in .13, .14, and .15 beta releases. What solved things for me was removing all `node_modules` directories in my turborepo setup and reinstalling all deps.

**Reply by jetaggart** [reply]
@chinanderm hmm that didn't seem to make a difference, although that makes me think it's a tooling issue (I use intellij). I did check it in vscode too and I see the same error though.

**Comment by sannajammeh** [comment]
Getting the same error in my codebase.

**Comment by KATT** [comment]
Maybe I was wrong on converting this issue, feel free to recreate it.

**Reply by KATT** [reply]
I haven't had time to really look into it yet

**Reply by jetaggart** [reply]
@KATT any luck on thinking through this? I'd like to be able to call the client directly in certain placed by just awaiting a call directly (for example in an auth flow). Do you have any suggestions/patterns on how to do this? I don't want to use the `useQuery` hook in certain places but simply await a response. I'm pretty stumped at this point.

I changed some of these to mutations, which allows me to do this:


```typescript
const meResult = await (trpc.me.get.useMutation()).mutateAsync()
```

which feels bad but works. However, I can't call this in a non-react context (redux thunk/inside of a context) since it's still a hook. I can set up a raw client but it feels like duplication.

**Reply by KATT** [reply, marked as answer]
tRPC is only accessible in the React tree, but you can set it to a global variable or `window` or something on first mount then use it elsewhere

Simplified:
```ts
window.trpc = trpc.useContext().client
```

Untested:
```tsx
export let trpcClient = null as unknown as ReturnType<typeof trpc.useContext>['client']
function MyApp() {
  trpcClient = useContext().client
}
```

**Reply by jetaggart** [reply]
That seems to work. I tested this and it works well. 

FWIW, `useContext` was also what I as looking for. I refactored my stuff to sit within the react tree and did this pattern which worked well:

```typescript
function Something() {
    const trpcContext = trpc.useContext()
    
    useEffect(() => {
      //..
      const thingResult = await trpcContext.thing.fetch()
    }, [someTrigger])
}
```

Thank you for all the help!

**Comment by TenchuBlossom** [comment]
Also having the same issue, would be great if we could get a fix for this :) trpc is awesome btw!

**Comment by aprilmintacpineda** [comment]
Actually the answer is in the homepage of trpc website.

```ts
export const trcpProxyClient = createTRPCProxyClient<AppRouter>({
  links: [trpcHttpBatchLink]
});

// then whenever we need it trcpProxyClient.<your-query-name-here>.<query | mutation>
const result = await trcpProxyClient.validateAuth.query();
```

This is going to be a separate client, so you can still use `react-query`.

**Reply by jetaggart** [reply]
You can do this but it is using a different client instance. Ideally, we are only configuring the client once and using that everywhere. I think `useContext().client` should be typed correctly, seems like a mistake that it's not.

**Reply by aprilmintacpineda** [reply]
Can you tell me more?

In this case I actually use the same `trpcHttpBatchLink` -- which is `httpBatchLink(...opts)` -- for both the proxy client and react-query client, I think the clients are all just using this `httpBatchLink` right?

**Reply by KATT** [reply]
https://github.com/trpc/trpc/discussions/2926#discussioncomment-5647033

**Reply by aprilmintacpineda** [reply]
I mean why is doing this not ideal?

**Reply by jetaggart** [reply]
Because you'd be configuring two clients, there's transformer arguments and I use quite a few links. It's not ideal but honestly not a huge deal imo.  The createClient is doing the same thing as what you're doing I suppose, and your option is better in this case. Just weird that it comes out untyped. 

p.s. I made a mistake in my earlier response. `useContext().client` is typed correctly, just `createClient` is not typed, which your solution is better since we'd just be wiring up everything anyways.

**Reply by silvio2402** [reply]
@KATT I know this is a rather old discussion, but is there a particular reason why we can't have a typed `trpc.createClient`? It would be great so we can just import the created `trpcClient` and use its `.query` or `.mutation` functions where we aren't in a React tree (e.g. in Redux `createApi`). Would this make sense? Can't think of a reason why it wouldn't so I need some closure 😉.
Edit: I suppose it's because of performance optimizations: #2716 is that the case?
